### ISSUE TITLE
smtp to grpc gateway

### ISSUE THREAD

**Issue by jsbucy** [envoyproxy/envoy#9133]
**Issue Template**

*Title*: smtp to grpc gateway

*Description*:
I'm interested in implementing an smtp-to-grpc gateway as an envoy plugin. The plugin would transcode rfc5321 to a grpc/protobuf intra-cluster representation "smtp.proto". This would open up access to any language with grpc support; you can receive smtp by writing a grpc handler. I'm thinking that this approach could be useful in situations where the business logic you want to implement is simple but the wire protocol is non-trivial and there aren't modular/resuable implementations for many languages.

SMTP does an in-band TLS upgrade via the STARTTLS command https://tools.ietf.org/html/rfc3207 and this is probably a hard requirement for most users. AIUI juggling the transport on the fly like this is not possible in envoy today so adding support for that may be a heavier lift than smtp itself. I don't know off the top of my head how many other protocols work that way.

**Comment by dio** [comment]
cc. @lizan

**Comment by lizan** [comment]
This should be doable within existing extension mechanism, with SMTP L4 filter + grpc bridge, it will be more generically reusable to split them to two.

> SMTP does an in-band TLS upgrade via the STARTTLS command https://tools.ietf.org/html/rfc3207 and this is probably a hard requirement for most users. AIUI juggling the transport on the fly like this is not possible in envoy today so adding support for that may be a heavier lift than smtp itself. I don't know off the top of my head how many other protocols work that way.

Not sure how exactly `STARTTLS` works, but if you're able to handle it within listener filter, then the transport socket can be kicked in after that.

**Comment by jsbucy** [comment]
In smtp, the client says EHLO, the server returns a list of capabilities and then the client says STARTTLS. But there is still a long tail of clients that don't do starttls so it needs to work both ways.

**Comment by lizan** [comment]
@jsbucy right, that's the purpose of the listener filter, you can return list of capabilities there and see whether client says `STARTTLS`, then select right filter chain (TLS or not) based on that. It will be similar to tls_inspector but more complicated though.

**Comment by jsbucy** [comment]
Ok so we would have 2 filters: the "smtp starttls" listener filter that just responds to ehlo and starttls and then the "smtp transcode" network filter that does all of the grpc stuff?

I think it is technically valid for a client to send one smtp transaction in the clear and then say STARTTLS and send another one which would break this implementation. I can't imagine why the client would do that but I don't know for sure that noone does.

**Comment by lizan** [comment]
> Ok so we would have 2 filters: the "smtp starttls" listener filter that just responds to ehlo and starttls and then the "smtp transcode" network filter that does all of the grpc stuff?

Sounds reasonable, "SMTP transcode" can be split to SMTP handling (like HTTP connection manager filter) and GRPC-SMTP L7 filter (like HTTP router filter).


> I think it is technically valid for a client to send one smtp transaction in the clear and then say STARTTLS and send another one which would break this implementation. I can't imagine why the client would do that but I don't know for sure that noone does.

At a quick glance on RFC 3207 I don't think that's valid in term of spec? I doubt anyone has use case on that, so we can ignore that case unless someone will use it.

**Comment by jsbucy** [comment]
In this design, you'd configure the smtp capabilities list into the envoy plugin. The listener filter would
- always return a 2xx success initial response (smtp "greeting") unless envoy itself had some kind of internal error
- return the same set of capabilities to every client
- always initiate tls in response to STARTTLS unless there was some kind of internal error

If it were possible to switch the transport socket on the fly, this could all be directed by the upstream/grpc backend which might want to customize behavior for different clients. Just for the sake of argument, what would it take to implement that in envoy? I've heard some discussion offline that it might not be a very extensive change?

I think we've identified a workable design here but just thought I'd ask. I don't know if anyone else could take advantage of this feature so I wouldn't put it on the critical path here unless it's straightforward and uncontroversial.

**Comment by lizan** [comment]
> If it were possible to switch the transport socket on the fly, this could all be directed by the upstream/grpc backend which might want to customize behavior for different clients. Just for the sake of argument, what would it take to implement that in envoy? I've heard some discussion offline that it might not be a very extensive change?

Let's open a new issue for this topic, this is not impossible but it is heavy (not only implementation, and tests), basically you want to shutdown existing filter chain and use existing connection to instantiate another filter chain. I'm not sure how useful this is but worth discussion. @envoyproxy/maintainers WDYT?

**Comment by jsbucy** [comment]
So after staring at the TLS Inspector filter code I'm not sure that approach will work for smtp. The gist of how tls inspector works is it accumulates the TLS ClientHello message in the socket buffer (with MSG_PEEK) until we can "dry fire" SSL_do_handshake on it to parse out the SNI and ALPN.

In the case of smtp, the client will say STARTTLS immediately after EHLO say 90% of the time but I'm not sure how to make this work when the client doesn't say starttls. In that case, we need to send ehlo to the upstream (presumably with the saved argument the client sent us) and wait for the response before passing the next command. But the listener doesn't have access to the upstream connection? Judging based on the comments at the beginning of onRead:

>  This receive code is somewhat complicated, because it must be done as a MSG_PEEK because  there is no way for a listener-filter to pass payload data to the ConnectionImpl and filters that get created later.

It seems like another approach would be to write a wrapper for TransportSocket that is more "in the middle" of the bytestream that would watch the smtp session for STARTTLS or until the session changes out of line/command mode (auth/data/bdat) at which point it becomes a pass-through.

**Comment by jsbucy** [comment]
Blocked on starttls #9577 

It seems like just getting envoy to upgrade all of your incoming smtp traffic to tls-wrapped for intra-cluster would be a win since you'd be terminating tls from the public internet with boringssl in envoy.

**Comment by jsbucy** [comment]
I'm working on a PR for STARTTLS

**Comment by jsbucy** [comment]
some progress on starttls on #9577
